src/middlewared/middlewared/plugins/account.py

```diff
@@ -27,50 +81,477 @@ class UserService(CRUDService):
+
+            await self.__set_groups(pk, groups)
+
+        except Exception:
+            if pk is not None:
+                await self.middleware.call('datastore.delete', 'account.bsdusers', pk)
+            if new_homedir:
+                # Be as atomic as possible when creating the user if
+                # commands failed to execute cleanly.
+                shutil.rmtree(data['home'])
+            raise
+
+        await self.middleware.call('service.reload', 'user')
+
+        await self.__set_smbpasswd(data['username'], password)
+        return pk
+
+    @accepts(
+        Int('id'),
+        Patch(
+            'user_create',
+            'user_update',
+            ('attr', {'update': True}),
+            ('rm', {'name': 'group_create'}),
+        ),
+    )
+    async def do_update(self, pk, data):
+
+        user = await self._get_instance(pk)
+
+        verrors = ValidationErrors()
+
+        if 'group' in data:
+            group = await self.middleware.call('datastore.query', 'account.bsdgroups', [('id', '=', data['group'])])
+            if not group:
+                verrors.add('group', f'Group {data["group"]} not found', errno.ENOENT)
+            group = group[0]
+        else:
+            group = user['group']
+            user['group'] = group['id']
+
+        await self.__common_validation(verrors, data, pk=pk)
+
+        home = data.get('home') or user['home']
+        if data.get('sshpubkey') and not home.startswith('/mnt'):
+            verrors.add('sshpubkey', 'Home directory is not writable, leave this blank"')
+
+        # Do not allow attributes to be changed for builtin user
+        if user['builtin']:
+            for i in ('group', 'home', 'home_mode', 'uid', 'username'):
+                if i in data:
+                    verrors.add(i, 'This attribute cannot be changed')
+
+        if verrors:
+            raise verrors
+
+        # Copy the home directory if it changed
+        if 'home' in data and data['home'] not in (user['home'], '/nonexistent'):
+            home_copy = True
+            home_old = user['home']
+        else:
+            home_copy = False
+
+        user.update(data)
+
+        password = await self.__set_password(user)
+
+        await self.__update_sshpubkey(user, group['bsdgrp_group'])
+
+        home_mode = user.pop('home_mode', None)
+        if home_mode is not None:
+            if not user['builtin'] and os.path.exists(user['home']):
                 try:
-                    with open(keysfile, 'r') as f:
-                        oldpubkey = f.read()
+                    os.chmod(user['home'], int(home_mode, 8))
+                except OSError:
+                    self.logger.warn('Failed to set homedir mode', exc_info=True)
+
+        if home_copy:
+            def do_home_copy():
+                subprocess.run(f"su - {user['username']} -c '/bin/cp -a {home_old}/* {user['home']}/'")
```

Comment:
We may wish to perform a check that new home directory is not nested under old home directory, otherwise this will be no-op (`cp: cannot copy a directory, 'test/a', into itself, 'test/a/a'`)